ols/src/prompts/prompt_generator.py

```diff
@@ -1,78 +1,126 @@
-    CONTEXT_PLACEHOLDER,
-    HISTORY_PLACEHOLDER,
-    QUERY_PLACEHOLDER,
     QUERY_SYSTEM_INSTRUCTION,
     USE_CONTEXT_INSTRUCTION,
     USE_HISTORY_INSTRUCTION,
 )
 
-PromptConfiguration = namedtuple("PromptConfiguration", "provider model")
-
-# This dictionary might be expanded in the future, if some combination of model+provider
-# requires specifically updated prompt.
-prompt_configurations = {
-    PromptConfiguration(PROVIDER_BAM, GRANITE_13B_CHAT_V2): QUERY_SYSTEM_INSTRUCTION,
-    PromptConfiguration(PROVIDER_OPENAI, GPT35_TURBO): QUERY_SYSTEM_INSTRUCTION,
-    PromptConfiguration(
-        PROVIDER_WATSONX, GRANITE_13B_CHAT_V2
-    ): QUERY_SYSTEM_INSTRUCTION,
-    PromptConfiguration(PROVIDER_AZURE_OPENAI, GPT35_TURBO): QUERY_SYSTEM_INSTRUCTION,
-}
-
-
-def generate_prompt(
-    provider: str,
-    model: str,
-    model_options: Optional[dict[str, Any]],
-    query: str,
-    history: list[str],
-    rag_context: str,
-) -> tuple[PromptTemplate, dict[str, Any]]:
-    """Dynamically creates prompt template and input values specification for LLM.
-
-    Prompt template is created by combining these values:
-    - provider
-    - model
-    - model-specific configuration
-    - RAG (if enabled and exists)
-    - history
-    - system prompt
-    - user query
-    """
-    # Add system instruction to the prompt first.
-    selector = PromptConfiguration(provider, model)
-    prompt = prompt_configurations.get(selector, QUERY_SYSTEM_INSTRUCTION)
-    llm_input_values = {"query": query}
-
-    # Add optional instruction to the prompt & create prompt inputs.
-    if len(rag_context) > 0:
-        prompt += USE_CONTEXT_INSTRUCTION
-        prompt += CONTEXT_PLACEHOLDER
-
-        llm_input_values["context"] = rag_context
-
-    if len(history) > 0:
-        prompt += USE_HISTORY_INSTRUCTION
-        prompt += HISTORY_PLACEHOLDER
-        llm_input_values["chat_history"] = "\n".join(history)
-
-    prompt += QUERY_PLACEHOLDER
-
-    return PromptTemplate.from_template(prompt), llm_input_values
+
+def restructure_rag_context_pre(text: str, model: str) -> str:
+    """Restructure rag text - pre truncation."""
+    if ModelFamily.GRANITE in model:
+        return "\n[End]\n[Document]\n" + text
+    else:
+        return "\n\nDocument:\n" + text
+
+
+def restructure_rag_context_post(text: str, model: str) -> str:
+    """Restructure rag text - post truncation."""
+    if ModelFamily.GRANITE in model:
+        return text.removeprefix("\n[End]") + "\n[End]"
+    else:
+        return "\n" + text.lstrip("\n") + "\n"
+
+
+def restructure_history(text: str, model: str) -> str:
+    """Restructure history."""
+    if text.startswith("human: "):
+        if ModelFamily.GRANITE in model:
+            return "\n<|user|>\n" + text.removeprefix("human: ")
+    else:
+        if ModelFamily.GRANITE in model:
+            return "\n<|assistant|>\n" + text.removeprefix("ai: ")
+
+    # No processing required here for gpt.
+    return text
```

Comment:
@asamal4 If the model is not Granite , we don't have to do any of this checks right , can we avoid those checks something like this ?
```
def restructure_history(text: str, model: str) -> str:
    """Restructure history."""
    # if model is not granite, then no processing required.
    if ModelFamily.GRANITE not in model:
        return text

    if text.startswith("human: "):
            return "\n<|user|>\n" + text.removeprefix("human: ")
    else:
        if text.startswith("ai: "):
            return "\n<|assistant|>\n" + text.removeprefix("ai: ")
    return text
```